### DISCUSSION TITLE
(ANSWERED) Adding Messier objects to a plot but they do not align with stars

### DISCUSSION THREAD

**Discussion by bathoorn** [skyfielders/python-skyfield#634] [Q&A] [answered]
I think that i am doing something wrong, but i cannot find what. So some help would be appreciated.

I am loading DSO data from stellarium/nebula/default/catalog.txt with the following code.
I think the ra/dec is in J2000 but i am not 100% sure (T0 from constants is that J2000?). 

```
from skyfield.constants import T0

_COLUMN_NAMES = (
    'DSOID', 'RAdeg', 'DEdeg', 'Bmag', 'Vmag',
    'OType', 'MType', 'MajRarcmin', 'MinRarcmin', 'OAdegrees', 'RS', 'RSerror',
    'Plx', 'Plxerror', 'NRSdist', 'NRSdisterror', 'NGC', 'IC', 'M', 'C', 'B',
    'Sh2', 'VdB', 'RCW', 'LDN', 'LBN', 'Cr', 'Mel', 'PGC', 'UGC', 'Ced', 'Arp', 'VV', 'PK', 'PN',
    'SNR', 'ACO', 'HCG', 'ESO', 'VdBH', 'DWB', 'Tr', 'St', 'Ru', 'VdB-Ha',
)


def load_dataframe(fobj):

    try:
        from pandas import read_csv
    except ImportError:
        raise ImportError("NO PANDAS NO CANDO")

    fobj.seek(0)
    magic = fobj.read(2)
    compression = 'gzip' if (magic == b'\x1f\x8b') else None
    fobj.seek(0)

    df = read_csv(
        fobj, sep='	', names=_COLUMN_NAMES, compression=compression,
        comment='#',
        usecols=['DSOID', 'Vmag', 'RAdeg', 'DEdeg', 'M'],
        na_values=[''],
    )
    df.columns = (
        'dso_id', 'magnitude', 'ra_degrees', 'dec_degrees',
        'messier_id', 'NGC_id',
    )
    df = df.assign(
        ra_hours = df['ra_degrees'] / 15.0,
        epoch_year = T0,
    )
    df.loc[df['messier_id'] != 0.0, 'label'] = 'M' + df['messier_id'].astype(str).str.slice(0, -2, 1)
    df = df[~df.messier_id.eq(0.0)]
    return df.set_index('dso_id')
```

Using this code i can load stars and messier objects using the following:

```
# The Hipparcos mission provides our star catalog.

with load.open(hipparcos.URL) as f:
    stars = hipparcos.load_dataframe(f)

# DSO's from stellarium

with open('catalog.txt') as f:
    dsolist = dsos.load_dataframe(f)
```

But when i try to plot them in the same image they do not align.
Am i doing something wrong? Is the data in a different format than i think it is?

Here is an example images so you can see what i mean.

![summer-triangle](https://user-images.githubusercontent.com/760615/132321625-42940dc9-bcd9-4dbb-b951-53cbb36add4f.png)

**Comment by bathoorn** [comment, marked as answer]
Lots of mistakes in the dso dataframe code.

This is what made it work

Now it looks like this:

![summer-triangle](https://user-images.githubusercontent.com/760615/132340854-4127599e-312c-46da-96e1-c4b7f11b5daf.png)


```
from skyfield.constants import T0

_COLUMN_NAMES = (
    'DSOID', 'RAdeg', 'DEdeg', 'Bmag', 'Vmag',
    'OType', 'MType', 'MajRarcmin', 'MinRarcmin', 'OAdegrees', 'RS', 'RSerror',
    'Plx', 'Plxerror', 'NRSdist', 'NRSdisterror', 'NGC', 'IC', 'M', 'C', 'B',
    'Sh2', 'VdB', 'RCW', 'LDN', 'LBN', 'Cr', 'Mel', 'PGC', 'UGC', 'Ced', 'Arp', 'VV', 'PK', 'PN',
    'SNR', 'ACO', 'HCG', 'ESO', 'VdBH', 'DWB', 'Tr', 'St', 'Ru', 'VdB-Ha',
)


def load_dataframe(fobj):
    """Given an open file for `catalog.txt`, return a parsed dataframe.

    If your copy of ``catalog.txt`` has already been unzipped, pass the
    optional argument ``compression=None``.

    """
    try:
        from pandas import read_csv, set_option
    except ImportError:
        raise ImportError("NO PANDAS NO CANDO")

    fobj.seek(0)
    magic = fobj.read(2)
    compression = 'gzip' if (magic == b'\x1f\x8b') else None
    fobj.seek(0)

    df = read_csv(
        fobj, sep='	', names=_COLUMN_NAMES, compression=compression,
        comment='#',
        usecols=['DSOID', 'RAdeg', 'DEdeg', 'Bmag', 'Vmag', 'M'],
        na_values=[''],
    )
    df.columns = (
        'dso_id', 'ra_degrees', 'dec_degrees', 'magnitudeB', 'magnitude',
        'messier_id',
    )
    df = df.assign(
        ra_hours = df['ra_degrees'] / 15.0,
        epoch_year = 2000.0,
    )
    df.loc[df['messier_id'] != 0.0, 'label'] = 'M' + df['messier_id'].astype(str)
    df = df[~df.messier_id.eq(0.0)]
    return df.set_index('dso_id')
```

**Reply by brandon-rhodes** [reply]
The only mistake I caught in your original code during a quick read-through was `epoch_year = T0`, then noticed that you had already answered your own question. I like how your plot works now and am glad the problem is fixed and you can move on!

**Reply by bathoorn** [reply]
Yeah! this project is cool. I like how it is very open. And you can basically add other datasets to it.
I sort of abused the Star object right now to add DSO's but maybe in the future i will add something for that.  For now the Star object has everything in need.

**Reply by bathoorn** [reply]
But if i use 2000.0 for the epoch_year that should be good right?

So for the dso data i don't have any information such as ra_mas_per_year and dec_mas_per_year will that make it less accurate over time?

**Reply by brandon-rhodes** [reply]
> I sort of abused the Star object …

The name “star” was simply convenient — it’s really intended for any object with fixed (or only slowly changing) RA and dec, so your use of it here is fine!

> But if i use 2000.0 for the epoch_year that should be good right?

Yes! It simply needs to be a numeric year, rather than a Julian Date.

> So for the dso data i don't have any information such as ra_mas_per_year and dec_mas_per_year will that make it less accurate over time?

Yes, if the nebula or star cluster is moving across the sky over very long periods of time. My guess would be that M objects tend to be very distant compared to many common stars, though, and that they might not have a large enough proper motion to make any difference over short periods like a few thousand years?
